prusti-tests/tests/verify_overflow/pass/competitive_programming/dp-on-trees/MaxCoinsInBinaryTree.rs

```diff
@@ -0,0 +1,217 @@
+impl Tree {
+    #[trusted]
+    #[pure]
+    #[ensures (result == self.coins)]
+    pub fn coins(&self) -> isize {
+        self.coins
+    }
+
+    #[trusted]
+    #[pure]
+    #[ensures (result == self.idx)]
+    #[ensures(result >= 0 && result < self.n)]
+    pub fn idx(&self) -> isize {
+        self.idx
+    }
+
+    #[trusted]
+    #[requires(i  >=  0  && i < nn)]
+    #[ensures (result.coins() == c)]
+    #[ensures (result.idx() == i)]
+    #[ensures(same_n(&result))]
+    pub fn new(
+        nn: isize,
+        i: isize,
+        c: isize,
+        l: Option<Box<Tree>>,
+        r: Option<Box<Tree>>,
+    ) -> Self {
+        Tree {
+            n: nn,
+            idx: i,
+            coins: c,
+            left: l,
+            right: r,
+        }
+    }
+}
+
+#[pure]
+pub fn same_n(node: &Tree) -> bool {
+    let mut result = true;
+    match &(*node).left {
+        None => {}
+        Some(box l) => {
+            result &= same_n(&l);
+            result &= (node.n == l.n);
+        }
+    }
+
+    match &(*node).right {
+        None => {}
+        Some(box r) => {
+            result &= same_n(&r);
+            result &= (node.n == r.n);
+        }
+    }
+    result
+}
+
+pub struct VecWrapperI32 {
+    _ghost_size: usize,
+    v: Vec<isize>,
+}
+
+impl VecWrapperI32 {
+    #[trusted]
+    #[pure]
+    #[ensures (0 <= result)]
+    pub fn len(&self) -> isize {
+        self._ghost_size as isize
+    }
+
+    #[trusted]
+    #[requires(size > 0)]
+    #[ensures (result.len() == size)]
+    #[ensures (forall(|i: isize| (0 <= i && i < result.len()) ==> result.lookup(i) == 0))]
+    pub fn new(size: isize) -> Self {
+        Self {
+            _ghost_size: size as usize,
+            v: vec![0; size as usize],
+        }
+    }
+
+    #[trusted]
+    #[pure]
+    #[requires (0 <= index && index < self.len())]
+    pub fn lookup(&self, index: isize) -> isize {
+        self.v[index as usize]
+    }
+
+    #[trusted]
+    #[requires(0 <= idx && idx < self.len())]
+    #[ensures(self.len() == old(self.len()))]
+    #[ensures(self.lookup(idx) == value)]
+    #[ensures(forall(|i: isize|
+        (0 <= i && i < self.len() && i != idx) ==>
+        self.lookup(i) == old(self.lookup(i))))]
+    pub fn set(&mut self, idx: isize, value: isize) -> () {
+        self.v[idx as usize] = value
+    }
+}
+
+#[pure]
+#[ensures (result >= a && result >= b)]
+#[ensures (result == a || result == b)]
+fn max(a: isize, b: isize) -> isize {
+    if a > b {
+        a
+    } else {
+        b
+    }
+}
+
+#[trusted]
+#[pure]
+#[ensures(result == a + b)]
+fn add(a: isize, b: isize) -> isize {
+    a + b
+}
+
+#[pure]
+fn solve1_rec(node: &Tree) -> isize {
+    let mut result = node.coins;
+    match &(*node).left {
+        None => {}
+        Some(box l) => {
+            result = add(result, solve2_rec(&l));
+        }
+    }
+
+    match &(*node).right {
+        None => {}
+        Some(box r) => {
+            result = add(result, solve2_rec(&r));
```

Comment:
I have looked into the generated Viper code. The error reported on the Viper level is that the return value of `solve2_rec` is not guaranteed to be in the range of `isize`, which is clearly a bug in either Prusti or Viper.

Reply:
Could it be related to https://github.com/viperproject/silicon/issues/367?

Reply:
> Could it be related to [viperproject/silicon#367](https://github.com/viperproject/silicon/issues/367)?

I do not see from where we could get non-linear arithmetic in this example, so it is probably unrelated.